### PR TITLE
Support CREATE TABLE and DROP TABLE

### PR DESCRIPTION
**Pull request by mrzzzrm**
The SQLParser only supports DataTypes INT (which I translate to LONG right now), DOUBLE and TEXT, so we should probably expand that. No way to specify column nullability, Table MVCC usage, column encoding, thanks to the sqlparser.

### CODE REVIEW

**Comment by mrks** [comment]
That was fast!

> No way to specify column nullability

Why not? SQL has "NULL/NOT NULL"

**Comment by mrzzzrm** [comment]
yeah, but afaict the SQLParser doesn't support it

```c++
  // Represents definition of a table column
  struct ColumnDefinition {
    enum DataType {
      UNKNOWN,
      TEXT,
      INT,
      DOUBLE
    };

    ColumnDefinition(char* name, DataType type);
    virtual ~ColumnDefinition();

    char* name;
    DataType type;
  };
```

**Comment by mrzzzrm** [comment]
@mrks seems to work now.

File: src/lib/sql/sql_translator.cpp [hunk 1]
```diff
@@ -865,40 +867,79 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_show(const hsql::Show
...
+    }
+  }
+
+  return CreateViewNode::make(create_statement.tableName, std::make_shared<LQPView>(lqp, column_names));
+}
 
-      return CreateViewNode::make(create_statement.tableName, std::make_shared<LQPView>(lqp, column_names));
+std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_create_table(const hsql::CreateStatement& create_statement) {
+  AssertInput(!create_statement.select, "CREATE TABLE ... (SELECT...) not supported");
+  AssertInput(create_statement.columns, "CREATE TABLE: No columns specified. Parser bug?");
```
Comment (mrks) [inline src/lib/sql/sql_translator.cpp]: If you don't expect this to happen, shouldn't this be a non-input assertion?

File: src/lib/sql/sql_translator.cpp [hunk 2]
```diff
@@ -865,40 +867,79 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_show(const hsql::Show
...
+
+  auto column_definitions = TableColumnDefinitions{create_statement.columns->size()};
+
+  for (auto column_id = ColumnID{0}; column_id < create_statement.columns->size(); ++column_id) {
+    const auto* parser_column_definition = create_statement.columns->at(column_id);
+    auto& column_definition = column_definitions[column_id];
+
+    // TODO(anybody) SQLParser is missing support for Hyrises other types
```
Comment (mrks) [inline src/lib/sql/sql_translator.cpp]: "Hyrise's"

File: src/test/operators/maintenance/create_table_test.cpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+TEST_F(CreateTableTest, NameAndDescription) {
+  EXPECT_EQ(create_table->name(), "Create Table");
+  EXPECT_EQ(create_table->description(DescriptionMode::SingleLine),
+            "Create Table 't' ('a' int NON-NULL, 'b' float NULLABLE)");
+  EXPECT_EQ(create_table->description(DescriptionMode::MultiLine),
+            "Create Table 't' ('a' int NON-NULL\n'b' float NULLABLE)");
+}
+
+TEST_F(CreateTableTest, Execute) {
```
Comment (mrks) [inline src/test/operators/maintenance/create_table_test.cpp]: Could you please test the edge cases (Creating table that already exists, dropping table that doesn't)?

File: src/lib/logical_query_plan/create_table_node.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+std::string CreateTableNode::description() const {
+  std::ostringstream stream;
+
+  stream << "[CreateTable] Name: '" << table_name << "' (";
+  for (auto column_id = ColumnID{0}; column_id < column_definitions.size(); ++column_id) {
+    const auto& column_definition = column_definitions[column_id];
+
+    stream << "'" << column_definition.name << "' " << data_type_to_string.left.at(column_definition.data_type) << " ";
+    if (column_definition.nullable) {
+      stream << "NULLABLE";
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/logical_query_plan/create_table_node.cpp]: sure, whatever you prefer

File: src/lib/logical_query_plan/drop_table_node.cpp [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#include "drop_table_node.hpp"
+
+namespace opossum {
+
+DropTableNode::DropTableNode(const std::string& table_name)
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/logical_query_plan/drop_table_node.cpp]: Tracking these in #1189 now

File: src/lib/operators/insert.cpp [hunk 6]
```diff
@@ -57,15 +57,9 @@ class TypedSegmentProcessor : public AbstractTypedSegmentProcessor {
...
-        // Values to insert contain null, copy them
-        if (target_is_nullable) {
-          std::copy_n(casted_source->null_values().begin() + source_start_index, length,
-                      casted_target->null_values().begin() + target_start_index);
-        } else {
-          for (const auto null_value : casted_source->null_values()) {
-            Assert(!null_value, "Trying to insert NULL into non-NULL segment");
-          }
-        }
+        Assert(target_is_nullable, "Trying to insert NULL into non-NULL segment");
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/operators/insert.cpp]: good catch, I broke that.

File: src/lib/operators/maintenance/drop_table.cpp [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    : AbstractReadOnlyOperator(OperatorType::DropTable), table_name(table_name) {}
+
+const std::string DropTable::name() const { return "Drop Table"; }
+
+const std::string DropTable::description(DescriptionMode description_mode) const {
+  return name() + " '" + table_name + "'";
+}
+
+std::shared_ptr<const Table> DropTable::_on_execute() {
+  StorageManager::get().drop_table(table_name);
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/operators/maintenance/drop_table.cpp]: This would set a precedent. So far we're not using AssertInput in operators and we had a long discussion about why we probably do not want to survive exceptions in Operators (Worker exception forwarding, etc). So I think the best we can do is some sort of pre-execution sanity checking that the console can trigger

[the remark this answers is not in the document]
Comment (mrks) [reply src/lib/operators/maintenance/drop_table.cpp]: Ah, I forgot about that. I think we should have that sanity check, yes.

[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/operators/maintenance/drop_table.cpp]: Tracked in #1189 now
